File: search_alg_2.py
import sys

only_exact = False
if len(sys.argv) > 1:
    only_exact = sys.argv[1]

def evaluate_score(window, query):
    score = 0
    for i in range(0, len(query)):
        char_score = 3 if window[i] == query[i].encode("ascii") else -3
        score += char_score
    return score

data_file = open('./Data/data.txt', 'rb')
match_file = open('./Data/match_2.txt', "w")

query = list("CACCA")

moving_window = []
cumsums = []

while 1:

    chars = data_file.read(len(query))
    if len(chars) != len(query):
        break

    window = [elem.encode("ascii") for elem in list(chars.decode("ascii"))]
    score = evaluate_score(window, query)
    if score >= 3:
        if (only_exact and score == 3 * len(query)) or not only_exact:
            cur_pos = data_file.tell()
            match_file.write(str(cur_pos) + "\t" + chars.decode('ascii') + "\t" + str(score) + "\n")

    data_file.seek(1 - len(query), 1)

data_file.close()
match_file.close()

# The algorithm finds way more matches the alg. 1

File: search_alg_2.cpp
#include <iostream>
#include <fstream>
#include <deque>

const static std::string path = "./Data/";
const static std::string match_fname = "match_2_cpp.txt";

int32_t evaluate_score(std::string window, std::string query) {
    int32_t score = 0;
    for (size_t i = 0; i < query.size(); i++) {
        int32_t char_score = window[i] == query[i] ? 3 : -3;
        score += char_score;
    }
    return score;
}

int main(int argc, char ** argv) {

    std::string fname = "data.txt";
    std::string query = "CACCA";
    bool only_exact = false;

    if(argc > 1){
        query = std::string(argv[1]);
    }
    if(argc > 2) {
        fname = std::string(argv[2]) + ".txt";
    }
    if(argc > 3) {
        only_exact = std::stoi(argv[3]);
    }
    
    std::cout << "Looking for query " + query + " in file " + fname << std::endl;

    std::ifstream data_file(path + fname);
    std::ofstream match_file(path + match_fname);

    std::string moving_window(query.length(), 'x');                                           
    int32_t read_chars = 0;

    while(data_file.good()) {
        data_file.read(&moving_window[0], query.length());

        read_chars += query.length();
        if (read_chars % 100000 == 0) {
            std::cout << "Read " + std::to_string(read_chars) << std::endl;
        }

        int32_t score = evaluate_score(moving_window, query);
        if(score >= 3) {
            if((only_exact && score == 3 * query.length()) || !only_exact) {
                uint32_t cur_pos = data_file.tellg();
                match_file << cur_pos << "\t" + moving_window + "\t" << score << std::endl;
            }
        }
        
        data_file.seekg(1 - query.length(), std::ios::cur);

    }
    

    data_file.close();
    match_file.close();

    return 0;

}